mono/utils/flags.c

```diff
@@ -0,0 +1,217 @@
+} FlagData;
+
+static FlagData flag_meta[] = {
+#define DEFINE_FLAG_FULL(flag_type, ctype, c_name, cmd_name, def_value, comment) \
+	{ flag_type, &mono_##c_name, cmd_name },
+#define DEFINE_FLAG_READONLY(flag_type, ctype, c_name, cmd_name, def_value, comment) \
+	{ flag_type, NULL, cmd_name },
+#include "flag-definitions.h"
+};
+
+static const char*
+flag_type_to_str (MonoFlagType type)
+{
+	switch (type) {
+	case MONO_FLAG_BOOL:
+		return "bool";
+	case MONO_FLAG_BOOL_READONLY:
+		return "bool (read-only)";
+	case MONO_FLAG_INT:
+		return "int";
+	case MONO_FLAG_STRING:
+		return "string";
+	default:
+		g_assert_not_reached ();
+		return NULL;
+	}
+}
+
+static char *
+flag_value_to_str (MonoFlagType type, gconstpointer addr)
+{
+	switch (type) {
+	case MONO_FLAG_BOOL:
+	case MONO_FLAG_BOOL_READONLY:
+		return *(gboolean*)addr ? g_strdup ("true") : g_strdup ("false");
+	case MONO_FLAG_INT:
+		return g_strdup_printf ("%d", *(int*)addr);
+	case MONO_FLAG_STRING:
+		return *(char**)addr ? g_strdup_printf ("%s", *(char**)addr) : g_strdup ("\"\"");
+	default:
+		g_assert_not_reached ();
+		return NULL;
+	}
+}
+
+void
+mono_flags_print_usage (void)
+{
+#define DEFINE_FLAG_FULL(flag_type, ctype, c_name, cmd_name, def_value, comment) do { \
+		char *val = flag_value_to_str (flag_type, &mono_##c_name); \
+		g_printf ("  --%s (%s)\n\ttype: %s  default: %s\n", cmd_name, comment, flag_type_to_str (flag_type), val); \
+		g_free (val); \
+	} while (0);
+#include "flag-definitions.h"
+}
+
+/*
+ * mono_flags_parse_options:
+ *
+ *   Set flags based on the command line arguments in ARGV/ARGC.
+ * Remove processed arguments from ARGV and set *OUT_ARGC to the
+ * number of remaining arguments.
+ */
+void
+mono_flags_parse_options (const char **argv, int argc, int *out_argc, MonoError *error)
+{
+	int aindex = 0;
+	int i;
+	GHashTable *flag_hash = NULL;
+
+	while (aindex < argc) {
+		const char *arg = argv [aindex];
+
+		if (!(arg [0] == '-' && arg [1] == '-')) {
+			aindex ++;
+			continue;
+		}
+		arg = arg + 2;
+
+		if (flag_hash == NULL) {
+			/* Compute a hash to avoid n^2 behavior */
+			flag_hash = g_hash_table_new (g_str_hash, g_str_equal);
+			for (i = 0; i < G_N_ELEMENTS (flag_meta); ++i) {
+				g_hash_table_insert (flag_hash, (gpointer)flag_meta [i].cmd_name, &flag_meta [i]);
+			}
+		}
+
+		/* Compute flag name */
+		char *flagname = g_strdup (arg);
+		int len = strlen (arg);
+		int equals_sign_index = -1;
+		if (flagname [0] == 'n' && flagname [1] == 'o' && flagname [2] == '-') {
+			flagname += 3;
+		} else {
+			for (int i = 0; i < len; ++i) {
+				if (flagname [i] == '=') {
+					equals_sign_index = i;
+					flagname [i] = '\0';
+					break;
+				}
+			}
+		}
+
+		FlagData *flag = (FlagData*)g_hash_table_lookup (flag_hash, flagname);
+		if (!flag) {
+			aindex ++;
+			continue;
+		}
+
+		switch (flag->flag_type) {
+		case MONO_FLAG_BOOL:
+		case MONO_FLAG_BOOL_READONLY: {
+			gboolean negate = FALSE;
+			if (!strcmp (flag->cmd_name, arg)) {
```

Comment:
Store the length in the FlagData and check that first to avoid full string comparison?